File: mavigobi/models_regression/3-Polynomial_regression/polynomial_regression.R
# POLYNOMIAL REGREESION MODEL
# R - mavigobi


# PRE PROCESSING DATA

# Import dataset
dataset = read.csv('Position_Salaries.csv')
# Not define  X and y variables
dataset = dataset[,2:3]

# Overfitting problem. 
# Import packages
# install.packages("caTools")
# library(caTools) # import library
# set.seed(123) # define seed
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)

# Variables scaling
# It's important remember that Dummy variables are not numbers because Dummy are factors.
# training_set[, 2:3] = scale(training_set[,2:3]) # not : because R assume all rows and columns
# testing_set[,2:3] = scale(testing_set[,2:3])


# LINEAR REGRESSION MODEL

l_regression = lm(formula = Salary ~ ., 
                  data = dataset)
# summary(l_regression)


# POLYNOMIAL REGRESSION MODEL
# add as many columns as degrees of the polynomial
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_regression = lm(formula = Salary ~ .,
                     data = dataset)
# summary(poly_regression)


# DISPLAY OF LINEAR REGRESSION MODEL
library("ggplot2")
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = "blue") +
  geom_line(aes(x = dataset$Level, y = predict(l_regression, newdata = dataset)),
            color = "red") +
  ggtitle("Modelo de Regresión Lineal") +
  xlab("Nivel del empleado")
  ylab("Salario anual del empleado")
  
# DISPLAY OF POLYNOMIAL REGRESSION MODEL
  ggplot() +
    geom_point(aes(x = dataset$Level, y = dataset$Salary),
               color = "blue")+
    geom_line(aes(x = dataset$Level, y = predict(poly_regression, newdata = dataset)),
              color = "red")+
    ggtitle("Modelo de Regresión Polinómica")+
    xlab("Nivel del empleado") +
    ylab("Salario anual del empleado")
  
  
  
# DISPLAY OF POLYNOMIAL REGRESSION MODEL (HIGH ACCURACY)
library(ggplot2)
# x_grid increase number of data
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
            color = "blue")+
  geom_line(aes(x = x_grid, y = predict(poly_regression, newdata = data.frame(Level = x_grid,
                                                                                     Level2 = x_grid^2,
                                                                                     Level3 = x_grid^3,
                                                                                     Level4 = x_grid^4))),
            color = "red")+
  ggtitle("Modelo de Regresión Polinómica")+
  xlab("Nivel del empleado") +
  ylab("Salario anual del empleado")
  
  
# PREDICTION OF LINEAR REGRESSION MODEL
y_predict = predict(l_regression, newdata = data.frame(Level = 6.5))
  
# PREDICTION OF POLYNOMIAL REGRESSION MODEL
y_poly_predict = predict(poly_regression, newdata = data.frame(Level = 6.5,
                                                          Level2 = 6.5^2,
                                                          Level3 = 6.5^3,
                                                          Level4 = 6.5^4))

File: mavigobi/models_regression/3-Polynomial_regression/polynomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Wed Nov 22 21:22:41 2023

@author: mavigobi
"""

# POLYNOMIAL REGRESSION

# PRE PROCESSING DATA

# How to import  libraries 
import numpy as np;# Mathematical tool
import matplotlib.pyplot as plt; # Graphics tool
import pandas as pd; # Read data

# Import dataset
dataset = pd.read_csv('Position_Salaries.csv');
X = dataset.iloc[:, 1:2]
y = dataset.iloc[:,2]

# IN THIS CASE, THE VARIABLES SHOULDN'T BE DIVIDED

# Overfitting problem / Split data
# Dataset: training set & test set
"""
from sklearn.model_selection import train_test_split # four variables
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0) # It is common to sacrifice 20% of the data for the test
"""

# Variable scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)""" 

# LINEAR REGRESSION DATA
from sklearn.linear_model import LinearRegression
l_regression = LinearRegression()
l_regression.fit(X, y)

# POLYNOMIAL REGRESSION DATA
from sklearn.preprocessing import PolynomialFeatures
poly_regression = PolynomialFeatures(degree=4) # polynomial regression ax^2 + bx + c
X_poly = poly_regression.fit_transform(X) 
lin_regression = LinearRegression() # linear regression of polynomical data regression
lin_regression.fit(X_poly,y)

# DISPLAY OF LINEAR REGRESSION MODEL
plt.scatter(X, y, color = "blue")
plt.plot(X, l_regression.predict(X), color = "red")
plt.title("Modelo de Regresión Lineal")
plt.xlabel("Sueldo anual del empleado")
plt.ylabel("Nivel laboral del empleado")
plt.show()

# DISPLAY OF POLYNOMIAL REGRESSION MODEL
X_grid = np.arange(min(X["Level"]), max(X["Level"]), 0.01, dtype= float) 
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = "blue")
plt.plot(X_grid, lin_regression.predict(poly_regression.fit_transform(X_grid)), color = "red")
plt.title("Modelo de Regresión Polinómica")
plt.xlabel("Sueldo anual del empleado")
plt.ylabel("Nivel laboral del empleado")
plt.show()

# PREDICTION OF LINEAR REGRESSION AND POLYNOMIAL REGRESSION
l_regression.predict([[6.5]])
lin_regression.predict(poly_regression.fit_transform([[6.5]]))

